File: levelledmobs-plugin-bukkit/src/main/java/me/lokka30/levelledmobs/bukkit/listener/EntitySpawnListener.java
package me.lokka30.levelledmobs.bukkit.listener;

import me.lokka30.levelledmobs.bukkit.LevelledMobs;
import me.lokka30.levelledmobs.bukkit.api.data.keys.EntityKeyStore;
import me.lokka30.levelledmobs.bukkit.data.InternalEntityDataUtil;
import me.lokka30.levelledmobs.bukkit.logic.context.Context;
import org.bukkit.entity.LivingEntity;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.entity.EntitySpawnEvent;

public final class EntitySpawnListener extends ListenerWrapper {

    /*
    Constructors
     */

    public EntitySpawnListener() {
        super(true);
    }

    /*
    Methods
     */

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    public void handle(final EntitySpawnEvent event) {

        /*
        LevelledMobs only concerns LivingEntities
         */
        if(!(event.getEntity() instanceof final LivingEntity entity))
            return;

        /*
        Check if the entity has any non-persistent metadata to migrate
         */
        // wasSummoned
        var wasSummonedKeyStr = EntityKeyStore.WAS_SUMMONED.toString();
        InternalEntityDataUtil.setWasSummoned(
            entity,
            entity.hasMetadata(wasSummonedKeyStr) &&
                entity.getMetadata(wasSummonedKeyStr).stream().anyMatch(val -> val.asInt() == 1),
            true
        );
        entity.removeMetadata(wasSummonedKeyStr, LevelledMobs.getInstance());

        /*
        Fire the associated trigger.
         */
        LevelledMobs.getInstance().getLogicHandler().runFunctionsWithTriggers(
            new Context()
                .withEntity(entity)
                .withEntityType(entity.getType())
                .withWorld(entity.getWorld())
                .withLocation(entity.getLocation()),
            "on-entity-spawn", "on-mob-spawn"
        );
    }
}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/plugin/bukkit/listener/entity/EntitySpawnListener.kt
package io.github.arcaneplugins.levelledmobs.plugin.bukkit.listener.entity

import io.github.arcaneplugins.levelledmobs.plugin.bukkit.LevelledMobs.Companion.lmInstance
import io.github.arcaneplugins.levelledmobs.plugin.bukkit.listener.ListenerWrapper
import io.github.arcaneplugins.levelledmobs.plugin.bukkit.rule.component.context.Context
import io.github.arcaneplugins.levelledmobs.plugin.bukkit.rule.component.trigger.LmTrigger
import org.bukkit.entity.LivingEntity
import org.bukkit.event.EventHandler
import org.bukkit.event.entity.EntitySpawnEvent

//todo doc
class EntitySpawnListener : ListenerWrapper(imperative = true) {

    @EventHandler
    fun onEntitySpawn(event: EntitySpawnEvent) {
        /* handle context for this event */
        val context = Context(
            entity = event.entity,
            event = event,
            location = event.location,
            world = event.entity.world,
        )

        if (event.entity is LivingEntity)
            context.livingEntity = event.entity as LivingEntity

        /* handle triggers for this event */
        lmInstance.ruleManager.callRulesWithTrigger(
            trigger = LmTrigger.ON_ENTITY_SPAWN,
            context = context
        )
    }

}